Process_Library/GOCART2G_Process.F90

```diff
@@ -3347,72 +3345,27 @@ subroutine Aero_Compute_Diags (mie_table, km, klid, nbegin, nbins, rlow, rup, ch
 
 !  Initialize local variables
 !  --------------------------
-   nch = size(channels)
    i2 = size(rhoa,1)
    j2 = size(rhoa,2)
    allocate(fPMfm(nbins))
    allocate(fPM25(nbins))
 
 !  Get the wavelength indices
 !  --------------------------
-!  Must provide ilam550 for AOT calculation
-   allocate(wavelengths_index_profile(size(wavelengths_profile)))
-   allocate(wavelengths_index_vertint(size(wavelengths_vertint)))
-   wavelengths_index_profile = 0.
-   wavelengths_index_vertint = 0.
-   ilam550 = 1.
-   ilam470 = 0.
-   ilam870 = 0.
-   if(nch .gt. 1) then
-      do i = 1, nch
-         if ( channels(i) .ge. 5.49e-7 .and. &
-              channels(i) .le. 5.51e-7) ilam550 = i
-         if ( channels(i) .ge. 4.69e-7 .and. &
-              channels(i) .le. 4.71e-7) ilam470 = i
-         if ( channels(i) .ge. 8.69e-7 .and. &
-              channels(i) .le. 8.71e-7) ilam870 = i
-      enddo
-   endif
 
-   ! Channel values are 4.7e-7 5.5e-7 6.7e-7 8.7e-7 [meter]. Their indices are 1,2,3,4 respectively.
-   do i = 1, size(wavelengths_profile)
-      if ((wavelengths_profile(i) .ge. 5.49e-7) .and. (wavelengths_profile(i) .le. 5.51e-7)) then
-         wavelengths_index_profile(i) = 2.
-      else if ((wavelengths_profile(i) .ge. 4.69e-7) .and. (wavelengths_profile(i) .le. 4.71e-7)) then
-         wavelengths_index_profile(i) = 1.
-      else if ((wavelengths_profile(i) .ge. 6.69e-7) .and. (wavelengths_profile(i) .le. 6.71e-7)) then
-         wavelengths_index_profile(i) = 3.
-      else if ((wavelengths_profile(i) .ge. 8.68e-7) .and. (wavelengths_profile(i) .le. 8.71e-7)) then
-         wavelengths_index_profile(i) = 4.
-      else
-         print*,'wavelengths_profile of ',wavelengths_profile(i),' is an invalid value.'
-         return
-      end if
-   end do
+   ilam470 = mie%getChannel(4.70e-7)
+   if(ilam470 <= 0) ilam470 = 0
```

Comment:
This is an amazingly nice simplification in the logic.   Both in terms of length and clarity.

The perfectionist in me now wonders if we could even avoid the ugly 0 or 1 switches that control these and instead modify other logic to only use those wavelengths that are found rather than checking the (int) flags.  But that's beyond the scope of this project.